src/StreamJsonRpc/Reflection/RpcTargetInfo.cs

```diff
@@ -0,0 +1,685 @@
+                            await asyncDisposableTarget.DisposeAsync().ConfigureAwait(false);
+                        }
+                        else if (target is Microsoft.VisualStudio.Threading.IAsyncDisposable vsAsyncDisposableTarget)
+                        {
+                            await vsAsyncDisposableTarget.DisposeAsync().ConfigureAwait(false);
+                        }
+                        else if (target is IDisposable disposableTarget)
+                        {
+                            disposableTarget.Dispose();
+                        }
+                    }
+#pragma warning disable CA1031 // Do not catch general exception types
+                    catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+                    {
+                        if (exceptions is null)
+                        {
+                            exceptions = new List<Exception>();
+                        }
+
+                        exceptions.Add(ex);
+                    }
+                }
+
+                if (exceptions is object)
+                {
+                    if (exceptions.Count == 1)
+                    {
+                        ExceptionDispatchInfo.Capture(exceptions[0]).Throw();
+                    }
+
+                    throw new AggregateException(exceptions);
+                }
+            }
+        }
+
+        internal static MethodNameMap GetMethodNameMap(TypeInfo type)
+        {
+            MethodNameMap? map;
+            lock (MethodNameMaps)
+            {
+                if (MethodNameMaps.TryGetValue(type, out map))
+                {
+                    return map;
+                }
+            }
+
+            map = new MethodNameMap(type);
+
+            lock (MethodNameMaps)
+            {
+                if (MethodNameMaps.TryGetValue(type, out MethodNameMap? lostRaceMap))
+                {
+                    return lostRaceMap;
+                }
+
+                MethodNameMaps.Add(type, map);
+            }
+
+            return map;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="JsonRpcMethodAttribute"/> for a previously discovered RPC method, if there is one.
+        /// </summary>
+        /// <param name="methodName">The name of the method for which the attribute is sought.</param>
+        /// <param name="parameters">
+        /// The list of parameters found on the method, as they may be given to <see cref="JsonRpcRequest.TryGetTypedArguments(ReadOnlySpan{ParameterInfo}, Span{object})"/>.
+        /// Note this list may omit some special parameters such as a trailing <see cref="CancellationToken"/>.
+        /// </param>
+        internal JsonRpcMethodAttribute? GetJsonRpcMethodAttribute(string methodName, ReadOnlySpan<ParameterInfo> parameters)
+        {
+            if (this.targetRequestMethodToClrMethodMap.TryGetValue(methodName, out List<MethodSignatureAndTarget>? existingList))
```

Comment:
Do we need locks here or can a local rpc target object be added only before calling these?

Reply:
Oooh. good call. Accessing this dictionary without a lock would be dangerous. I'm adding the lock now.